File: codeforces/1600~1699/1612/e.rs
use std::io::{stdin, stdout, BufWriter, Write};
use std::str::FromStr;
use std::fmt::Debug;
use std::cmp::min;

fn read_int<T: FromStr>() -> T where <T as FromStr>::Err: Debug {
    let mut s = String::new();
    stdin().read_line(&mut s).unwrap();
    s.trim().parse::<T>().expect("invalid input")
}

fn read_vec<T: FromStr>() -> Vec<T> where <T as FromStr>::Err: Debug {
    let mut s = String::new();
    stdin().read_line(&mut s).unwrap();
    if s.is_empty() {
        return Vec::new();
    }
    s.split(" ").map(|x| x.trim().parse().expect("invalid input")).collect()
}

fn greater(x: (i64, i64), y: (i64, i64)) -> bool {
    x.0 * y.1 > x.1 * y.0
}

fn solve() {
    let stdout = stdout();
    let mut writer = BufWriter::new(stdout.lock());
    let n = read_int::<usize>();
    let mut m = vec![0; n];
    let mut k = vec![0; n];
    for i in 0..n {
        let row = read_vec::<i32>();
        m[i] = row[0];
        k[i] = row[1];
    }

    let max_k = *k.iter().max().unwrap();
    let max_m = *m.iter().max().unwrap();

    let mut best: (i64, i64) = (0, 1);
    let mut ans = Vec::new();
    for t in 1..=max_k {
        if t > max_m {
            break;
        }

        let mut score = vec![0; max_m as usize + 1];
        for i in 0..n {
            score[m[i] as usize] += min(t, k[i]);
        }

        let mut p = Vec::new();
        for i in 0..=max_m as usize {
            p.push((score[i], i));
        }

        p.sort();
        p.reverse();
        let mut total = 0;
        let mut list = Vec::new();
        for i in 0..t {
            total += p[i as usize].0 as i64;
            list.push(p[i as usize].1);
        }

        if greater((total, t as i64), best) {
            best = (total, t as i64);
            ans = list;
        }
    }

    writeln!(writer, "{}", ans.len()).ok();
    for i in 0..ans.len() {
        write!(writer, "{} ", ans[i]).ok();
    }
    writeln!(writer).ok();
}

fn main() {
    solve();
}

File: codeforces/1600~1699/1612/e.cc
#include <bits/stdc++.h>

using namespace std;

using i64 = long long;

// when t <= k[i]
// i contribute to answer 1
// when t > k[i]
// i contribute to answer k[i] / t

struct Fraction {
  i64 a, b;

  Fraction(i64 a, i64 b) : a(a), b(b) {}

  bool operator>(const Fraction& f2) {
    return a * f2.b > f2.a * b;
  }
};

void solve() {
  int n = 0;
  cin >> n;
  vector<int> m(n), k(n);
  for (int i = 0; i < n; ++i) {
    cin >> m[i] >> k[i];
  }

  int max_m = *max_element(m.begin(), m.end());
  int max_k = *max_element(k.begin(), k.end());
  Fraction best(0, 1);
  vector<int> ans;
  for (int t = 1; t <= max_k; ++t) {
    if (t > max_m) break;

    vector<int> score(max_m+1);
    for (int i = 0; i < n; ++i) {
      score[m[i]] += min(t, k[i]);
    }
    vector<pair<int,int>> p;
    for (int i = 0; i <= max_m; ++i) {
      p.emplace_back(score[i], i);
    }
    sort(p.rbegin(), p.rend());

    vector<int> list;
    int top = 0;
    for (int i = 0; i < t; ++i) {
      top += p[i].first;
      list.push_back(p[i].second);
    }
    Fraction f(top, t);
    if (f > best) {
      best = f;
      ans = list;
    }
  }

  cout << ans.size() << endl;
  for (int i : ans) cout << i << " ";
  cout << endl;
}

int main(void) {
  ios::sync_with_stdio(false);
  cin.tie(0);

  solve();
  return 0;
}
